### ISSUE TITLE
Rule proposal: Forbid certain values for component's props

### ISSUE THREAD

**Issue by Betree** [yannickcr/eslint-plugin-react#2725]
# Use case

In our current setup (based on StyledComponents/StyledSystem) we use aliases for font-sizes. For example:

```jsx
<Span fontSize="Paragraph" />

// Translated by styled-system to:
<Span fontSize="14px" />
```

We want to get away from this behavior for the reasons described [in this issue](https://github.com/opencollective/opencollective/issues/3109). But we want this transition to be smooth, so a good first step would be to emit simple warnings for the developers to change their habits for new developments.

## Rule Details

This rule checks all JSX elements and verifies that no forbidden value is passed to components through watched props. This rule is off by default.

Given the following configuration:

```json
"react/forbid-component-props-values": ["warning", {
  "forbid": [
    {
      "forbiddenFor": ["Span", "P", "H1", "H2", "H3"],
      "propName": ["fontSize", "lineHeight"],
      "forbiddenValue": ["Small", "Paragraph", "Caption"]
    }
  ]
}]
```

The following patterns are considered warnings:

```jsx
<Span fontSize="Small" />

<Span fontSize="Small" lineHeight="Small" />

<H1 lineHeight="Paragraph" />
```

The following patterns are **not** considered warnings:

```jsx
<Span fontSize="14px" />

<Span fontSize="14px" lineHeight="20px" />

<H1 lineHeight="14px" />
```

## Rule Options

```js
...
"react/forbid-component-props-values": ["warning", {
  "forbid": [
    {
      "forbiddenFor": [SomeComponent, AnotherComponent],
      "propName": "someProp",
      "forbiddenValue": "ForbiddenValue",
      "message": "Value $value cannot be used for $propName"
    }
  ]
}]
...
```

### `forbid`

An array specifying the forbidden values, as an object with the following keys:

- `forbiddenFor` (Component|Component[]): Component or list of components to watch
- `propName` (string|string[]): Prop or list of props to watch
- `forbiddenValue` (any|any[]): Value or list of values to forbidd 
- `message` (string): A message to display when the rule doesn't validate. `$value` and `$propName` placeholders will be replaced by their actual values. Defaults to `Value $value cannot be used for $propName`.

# Related rules

- [forbid-component-props](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-component-props.md)
- [forbid-dom-props](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-dom-props.md)

**Comment by ljharb** [comment]
I'm a bit confused. Typically, I'd suggest:
1. make propType warnings fail tests
1. wrap every propTypes object in https://npmjs.com/prop-types-exact so that passing extra props fails tests
1. add the new prop in a PR.
1. make a PR that migrates all existing uses to use the new prop.
1. make a PR that removes the old prop.

What's the point of using a linter rule here that can be overridden?

**Comment by Betree** [comment]
@ljharb For our use case, relying solely on prop types is not enough, we need to look at the value. The `fontSize` props for example should still accept strings. As per the example above:

```jsx
// Good
<Span fontSize="12px" />
<Span fontSize="2em" />

// Warning
<Span fontSize="Caption" />
```

We could achieve that through a custom proptypes validation function, but the warning would be at runtime only. The intent of bringing this to the linter is to have something that warns you directly in your editor, when you use the prop with a deprecated value.

**Comment by ljharb** [comment]
propTypes *do* look at the values, that's why they're superior to TS/Flow types.

And yes, propType warnings are at runtime only, but that's what tests are for :-)

**Comment by Betree** [comment]
We ended up making the migration in one go. Given our current setup, such feature could have been helpful to migrate progressively but is not needed anymore. Thanks for the input!
